src/components/teleporter.js

```diff
@@ -0,0 +1,377 @@
+    const { start, confirm, speed } = this.data;
+    const object3D = this.el.object3D;
+
+    if (!this.isTeleporting && userinput.get(start)) {
+      this.isTeleporting = true;
+      this.timeTeleporting = 0;
+      this.hit = false;
+      this.rayCurve.mesh.visible = true;
+      this.rayCurve.mesh.updateMatrixWorld();
+      this.rayCurve.mesh.material.opacity = MISS_OPACITY;
+      this.rayCurve.mesh.material.color.set(MISS_COLOR);
+      this.rayCurve.mesh.material.needsUpdate = true;
+      this.playSound("teleport-sound-start");
+    }
+
+    if (!this.isTeleporting) {
+      return;
+    }
+
+    if (userinput.get(confirm)) {
+      this.hitEntity.setAttribute("visible", false);
+      this.isTeleporting = false;
+      this.rayCurve.mesh.visible = false;
+
+      if (!this.hit || this.timeTeleporting < DRAW_TIME_MS) {
+        this.playSound("teleport-sound-cancel");
+        return;
+      }
+
+      this.teleport();
+      this.playSound("teleport-sound-stop");
+      return;
+    }
+
+    this.timeTeleporting += dt;
+    object3D.updateMatrixWorld();
+    object3D.matrixWorld.decompose(this.p0, q, vecHelper);
+    this.direction
+      .copy(FORWARD)
+      .applyQuaternion(q)
+      .normalize();
+    this.rayCurve.setDirection(this.direction);
+    this.v0.copy(this.direction).multiplyScalar(speed);
+
+    let collidedIndex = this.rayCurve.numPoints - 1;
+
+    this.hit = false;
+    this.parabola[0].copy(this.p0);
+    const timeSegment = 1 / (this.rayCurve.numPoints - 1);
+    for (let i = 1; i < this.rayCurve.numPoints; i++) {
+      const t = i * timeSegment;
+      parabolicCurve(this.p0, this.v0, t, vecHelper);
+      this.parabola[i].copy(vecHelper);
+
+      if (
+        checkLineIntersection(
+          this.parabola[i - 1],
+          this.parabola[i],
+          this.meshes,
+          this.raycaster,
+          LANDING_NORMAL,
+          MAX_LANDING_ANGLE,
+          this.hitPoint
+        )
+      ) {
+        this.hit = true;
+        collidedIndex = i;
+        break;
+      }
+    }
+
+    const percentToDraw = this.timeTeleporting > DRAW_TIME_MS ? 1 : this.timeTeleporting / DRAW_TIME_MS;
+    const percentRaycasted = collidedIndex / (this.rayCurve.numPoints - 1);
+    const segmentT = (percentToDraw * percentRaycasted) / (this.rayCurve.numPoints - 1);
+
+    for (let i = 0; i < this.rayCurve.numPoints; i++) {
+      const t = i * segmentT;
+      parabolicCurve(this.p0, this.v0, t, vecHelper);
+      this.rayCurve.setPoint(i, vecHelper);
+    }
+
+    const color = this.hit ? HIT_COLOR : MISS_COLOR;
+    const opacity = this.hit && this.timeTeleporting >= DRAW_TIME_MS ? HIT_OPACITY : MISS_OPACITY;
+    this.rayCurve.material.color.set(color);
+    this.rayCurve.material.opacity = opacity;
+
+    this.hitEntity.setAttribute("visible", this.hit);
+    if (this.hit) {
+      this.hitEntity.setAttribute("position", this.hitPoint);
+      this.hitEntity.object3D.traverse(o => {
+        o.matrixNeedsUpdate = true;
```

Comment:
This seems really surprising. The matrix updating code should handle updating children itself, you should only need to set it on hitEntity